### PR TITLE
Bugfix#344 report error when one of schema doesnt have the discriminator property

### PR DESCRIPTION
**Pull request by senthilkumarmohan**

### CODE REVIEW

**Comment by PerthCharern** [comment]
I don't think we need to add this validation in the reference resolver itself. In general, we try to avoid adding a validation into the reader and resolver unless we have to (e.g. YAML malformatting can't really be validated anywhere else).

This seems like it would fit better in the Validations Rules

[https://github.com/Microsoft/OpenAPI.NET/tree/master/src/Microsoft.OpenApi/Validations/Rules](https://github.com/Microsoft/OpenAPI.NET/tree/master/src/Microsoft.OpenApi/Validations/Rules)

This rule belongs to OpenApiSchemaRules.

File: test/Microsoft.OpenApi.Readers.Tests/V3Tests/OpenApiDiscriminatorTests.cs [hunk 1]
```diff
@@ -47,5 +51,51 @@ public void ParseBasicDiscriminatorShouldSucceed()
...
+                    new OpenApiValidatorError(nameof(OpenApiSchemaRules.ValidateAnyOfDiscriminator),
+                        "#/paths/~1pets~1{id}/get/responses/200/content/application~1json/schema/anyOf",
+                        string.Format(SRResource.Validation_SchemaDoesntContainDiscriminatorProperty, "pet2", "petType")),
+
+                    new OpenApiValidatorError(nameof(OpenApiSchemaRules.ValidateAnyOfDiscriminator), 
+                        "#/paths/~1pets~1{id}/get/responses/200/content/application~1json/schema/anyOf",
+                        string.Format(SRResource.Validation_SchemaDoesntContainDiscriminatorPropertyInRequiredFieldList, "pet2", "petType"))
+                });
+            }
+        }
```
Comment (PerthCharern) [inline test/Microsoft.OpenApi.Readers.Tests/V3Tests/OpenApiDiscriminatorTests.cs]: These tests belong to Microsoft.OpenApi.Tests/Validations. Could you move them there and adjust?

**Comment by PerthCharern** [comment]
Thanks for adding the tests to the reader. Could you help do the following?

1) In your tests, add checks to verify the final object in the DOM as well (instead of just verifying the errors)

2) Add some tests in the Microsoft.OpenApi.Tests/Validations/OpenApiSchemaValidationTests.cs

File: src/Microsoft.OpenApi/Properties/SRResource.resx [hunk 2]
```diff
@@ -210,6 +210,12 @@
   <data name="Validation_RuleAddTwice" xml:space="preserve">
     <value>The same rule cannot be in the same rule set twice.</value>
   </data>
+  <data name="Validation_SchemaDoesntContainDiscriminatorProperty" xml:space="preserve">
```
Comment (PerthCharern) [inline src/Microsoft.OpenApi/Properties/SRResource.resx]: >Validation_SchemaDoesntContainDiscriminatorProperty [](start = 14, length = 51)

Suggest renaming this to:

Validation_CompositeSchemaMustContainPropertySpecifiedInTheDiscriminator

File: src/Microsoft.OpenApi/Properties/SRResource.resx [hunk 3]
```diff
@@ -210,6 +210,12 @@
...
+    <value>Schema {0} doesn't contain discriminator property {1}.</value>
+  </data>
+  <data name="Validation_SchemaDoesntContainDiscriminatorPropertyInRequiredFieldList" xml:space="preserve">
```
Comment (PerthCharern) [inline src/Microsoft.OpenApi/Properties/SRResource.resx]: >Validation_SchemaDoesntContainDiscriminatorPropertyInRequiredFieldList [](start = 14, length = 70)

Suggest renaming this to:

Validation_CompositeSchemaRequiedFieldMustContainThePropertySpecifiedInTheDiscriminator

File: src/Microsoft.OpenApi/Properties/SRResource.resx [hunk 4]
```diff
@@ -210,6 +210,12 @@
   <data name="Validation_RuleAddTwice" xml:space="preserve">
     <value>The same rule cannot be in the same rule set twice.</value>
   </data>
+  <data name="Validation_SchemaDoesntContainDiscriminatorProperty" xml:space="preserve">
+    <value>Schema {0} doesn't contain discriminator property {1}.</value>
```
Comment (PerthCharern) [inline src/Microsoft.OpenApi/Properties/SRResource.resx]: >Schema {0} doesn't contain discriminator property {1}. [](start = 11, length = 54)

Composite Schema {0} must contain property specified in the discriminator {1}.

File: src/Microsoft.OpenApi/Properties/SRResource.resx [hunk 5]
```diff
@@ -210,6 +210,12 @@
...
+  </data>
+  <data name="Validation_SchemaDoesntContainDiscriminatorPropertyInRequiredFieldList" xml:space="preserve">
+    <value>Schema {0} doesn't contain discriminator property {1} in the required field list.</value>
```
Comment (PerthCharern) [inline src/Microsoft.OpenApi/Properties/SRResource.resx]: >Schema {0} doesn't contain discriminator property {1} in the required field list. [](start = 11, length = 81)

Composite schema {0} must contain property specified in the discriminator {1} in the required field list.

File: src/Microsoft.OpenApi/Validations/Rules/OpenApiSchemaRules.cs [hunk 6]
```diff
@@ -55,6 +57,72 @@ public static class OpenApiSchemaRules
...
+        /// Validates OneOf Discriminator
+        /// </summary>
+        public static ValidationRule<OpenApiSchema> ValidateOneOfDiscriminator =>
+            new ValidationRule<OpenApiSchema>(
+                (context, schema) =>
+                {
+                    // oneOf
+                    context.Enter("oneOf");
+
+                    if (schema.OneOf != null && schema.Discriminator != null)
```
Comment (PerthCharern) [inline src/Microsoft.OpenApi/Validations/Rules/OpenApiSchemaRules.cs]: The spec also says this:

"When using the discriminator, inline schemas will not be considered."

To be honest, this sentence is a bit ambiguous. It doesn't make sense to have an inline schema but the spec doesn't disallow it per se. 

For our purpose, I'd suggest we perform this check ONLY when the reference field is also present.

ie. Change the if condition to the version below that's a bit stricter:

`if (schema.OneOf != null && schema.Discriminator != null && schema.Reference != null)`

File: src/Microsoft.OpenApi/Validations/Rules/OpenApiSchemaRules.cs [hunk 7]
```diff
@@ -55,6 +57,72 @@ public static class OpenApiSchemaRules
...
+        /// <param name="context"></param>
+        /// <param name="ruleName"></param>
+        /// <param name="schemas">OneOf/AnyOf schemas</param>
+        /// <param name="discriminator">discriminator</param>
+        private static void ValidateSchemaListDiscriminator(IValidationContext context, string ruleName,
+                        IList<OpenApiSchema> schemas, OpenApiDiscriminator discriminator)
+        {
+            foreach (var schema in schemas)
+            {
+                if (!schema.Properties.ContainsKey(discriminator.PropertyName))
```
Comment (PerthCharern) [inline src/Microsoft.OpenApi/Validations/Rules/OpenApiSchemaRules.cs]: > if (!schema.Properties.ContainsKey(discriminator.PropertyName)) [](start = 15, length = 64)

Add

`&& schema.Reference != null`

File: src/Microsoft.OpenApi/Validations/Rules/OpenApiSchemaRules.cs [hunk 8]
```diff
@@ -55,6 +57,72 @@ public static class OpenApiSchemaRules
...
+                {
+                    context.AddError(new OpenApiValidatorError(ruleName, context.PathString,
+                            string.Format(SRResource.Validation_SchemaDoesntContainDiscriminatorProperty,
+                                        schema.Reference.Id, discriminator.PropertyName)));
+                }
+                if (!schema.Required.Contains(discriminator.PropertyName))
```
Comment (PerthCharern) [inline src/Microsoft.OpenApi/Validations/Rules/OpenApiSchemaRules.cs]: > if (!schema.Required.Contains(discriminator.PropertyName)) [](start = 15, length = 59)

Add

`&& schema.Reference != null`

File: src/Microsoft.OpenApi/Validations/Rules/OpenApiSchemaRules.cs [hunk 9]
```diff
@@ -55,6 +57,72 @@ public static class OpenApiSchemaRules
...
         // add more rule.
+
+        
+        /// <summary>
+        /// Checks if the schemas in the list contain a property with the property name specified by the discriminator.
+        /// </summary>
+        /// <param name="context"></param>
+        /// <param name="ruleName"></param>
+        /// <param name="schemas">OneOf/AnyOf schemas</param>
+        /// <param name="discriminator">discriminator</param>
```
Comment (PerthCharern) [inline src/Microsoft.OpenApi/Validations/Rules/OpenApiSchemaRules.cs]: Remove

**Comment by senthilkumarmohan** [comment]
Thanks for the review. I have moved test cases to OpenApiSchemaValidationTests (with changes) and updated code as per the comments.

File: src/Microsoft.OpenApi/Properties/SRResource.Designer.cs [hunk 10]
```diff
@@ -303,6 +303,25 @@ internal class SRResource {
...
+        internal static string Validation_CompositeSchemaMustContainPropertySpecifiedInTheDiscriminator {
+            get {
+                return ResourceManager.GetString("Validation_CompositeSchemaMustContainPropertySpecifiedInTheDiscriminator", resourceCulture);
+            }
+        }
+        
+        /// <summary>
+        ///   Looks up a localized string similar to Composite schema {0} must contain property specified in the discriminator {1} in the required field list..
+        /// </summary>
+        internal static string Validation_CompositeSchemaRequiedFieldMustContainThePropertySpecifiedInTheDiscriminator {
```
Comment (PerthCharern) [inline src/Microsoft.OpenApi/Properties/SRResource.Designer.cs]: >RequiedField [](start = 57, length = 12)

RequiredFieldList

File: src/Microsoft.OpenApi/Properties/SRResource.Designer.cs [hunk 11]
```diff
@@ -303,6 +303,25 @@ internal class SRResource {
...
+            get {
+                return ResourceManager.GetString("Validation_CompositeSchemaRequiedFieldMustContainThePropertySpecifiedInTheDiscrim" +
```
Comment (PerthCharern) [inline src/Microsoft.OpenApi/Properties/SRResource.Designer.cs]: >RequiedField [](start = 76, length = 12)

RequiredFieldList

Comment (PerthCharern) [inline src/Microsoft.OpenApi/Properties/SRResource.Designer.cs]: >Discrim [](start = 124, length = 7)

Discriminator

File: test/Microsoft.OpenApi.Readers.Tests/V3Tests/OpenApiDiscriminatorTests.cs [hunk 12]
```diff
@@ -1,12 +1,16 @@
...
+using System.Collections.Generic;
 using System.IO;
 using System.Linq;
 using FluentAssertions;
 using Microsoft.OpenApi.Models;
+using Microsoft.OpenApi.Properties;
 using Microsoft.OpenApi.Readers.ParseNodes;
 using Microsoft.OpenApi.Readers.V3;
+using Microsoft.OpenApi.Validations;
+using Microsoft.OpenApi.Validations.Rules;
```
Comment (PerthCharern) [inline test/Microsoft.OpenApi.Readers.Tests/V3Tests/OpenApiDiscriminatorTests.cs]: nit, can you revert changes in this file?

**Comment by PerthCharern** [comment]
Thank you for making this change. I have left a few minor comments. Once you have fixed that and merged in the latest master, we should be good to go :)

**Comment by PerthCharern** [comment]
This looks good now. I'll check it in.

Thanks a lot for your contribution :)

**Comment by senthilkumarmohan** [comment]
Thankyou very much

**Pull request merged**